### PR TITLE
Async thumbnail

### PR DESCRIPTION
**Pull request by hishamkaram**
This PR contain:
- add async_thumbnail task to generate thumbnails to improve layer upload time

### CODE REVIEW

**Comment by giohappy** [comment]
IMHO the celery_enabled check should be only done during application startup and the result stored, globally.  Otherwise a test connection is opened on each thumbnail creation.
Morever the "celery or thread" pattern could be generalized to be used elsewhere.

**Comment by giohappy** [comment]
I want to hear comments from others on this, but I'm concerned of the thread spawning. Without proper management (pools, etc.) we risk to have tons of threads running and pending in case of batch imports / updates.

And I think this pattern should in case be moved to a generic place (utils module or else) 

What's your opinion @afabiani?

**Comment by afabiani** [comment]
> I want to hear comments from others on this, but I'm concerned of the thread spawning. Without proper management (pools, etc.) we risk to have tons of threads running and pending in case of batch imports / updates.
> 
> And I think this pattern should in case be moved to a generic place (utils module or else)
> 
> What's your opinion @afabiani?

Yes, I agree.
For the time being if celery is not enabled I would keep the usual sync mechanism until we can spend some more time and resources to envisage a better and more stable threading mechanism.

**Comment by capooti** [comment]
Personally I'd avoid to generate thumbnails when changing styles for now, as it is time consuming and it can happen pretty often in the map composer. When needed the user can manually sync the thumbnail from the layer details view.

**Comment by hishamkaram** [comment]
@afabiani @capooti please take a look at the last commit

File: geonode/geoserver/tasks.py [hunk 1]
```diff
@@ -18,6 +18,8 @@
 #
 #########################################################################
 
+import threading
```
Comment (afabiani) [inline geonode/geoserver/tasks.py]: This most probably will cause pep8 formatting issues

**Comment by afabiani** [comment]
@hishamkaram

still pep8 issues
```
flake8 geonode
geonode/geoserver/signals.py:47:1: F811 redefinition of unused 'thumbnail_task' from line 24

```

The overall idea is good though. Thanks for your contribution.

**Comment by giohappy** [comment]
@hishamkaram please run flake8 before submitting new commits.

**Comment by afabiani** [comment]
Thanks @hishamkaram

**Comment by hishamkaram** [comment]
@afabiani you are welcom

**Comment by frafra** [comment]
This pull request contains a commit which introduces a regression: a18bbd4.

**Comment by hishamkaram** [comment]
@frafra please review the PR comments

**Comment by gannebamm** [comment]
> @frafra please review the PR comments

@hishamkaram what do you mean by that? Which comment should be reviewed and why? I am puzzled.

**Pull request merged**